> @oscarbenjamin `mypy` just bit me... No idea what to do since it is some completely other code compared to what I changed...

We're going to see things like this for some time I think. I'm not sure how to describe this but it's almost like mypy is not deterministic some how. Obviously it is deterministic but what exactly it picks up on can be quite sensitive to changes in the codebase. I don't fully understand how it works but I imagine that in mypy's mind the codebase is just a massive mess and sometimes it can pick out something it doesn't like but potentially a small change somewhere else might mean that it doesn't notice any more.

In the short term there is always a quick fix which is just to add `type:ignore` at the exact line that mypy refers to.

The particular error you are seeing is
```console
$ mypy sympy
sympy/core/function.py:2298: error: Incompatible types in assignment (expression has type "Callable[[Any, KwArg(Any)], Any]", base class "EvalfMixin" defined the type as "Callable[[Any, Any, Any, Any, Any, Any, Any], Any]")
Found 1 error in 1 file (checked 1440 source files)
```
The line that it complains about is the assignment `n = evalf` below:
https://github.com/sympy/sympy/blob/e451705e52ed22ffae88da0cbf685b5facb3b7a2/sympy/core/function.py#L2295-L2298
Really the problem though is the fact that the `evalf` method here does not have a consistent signature with the evalf method in the `EvalfMixin` superclass:
https://github.com/sympy/sympy/blob/e451705e52ed22ffae88da0cbf685b5facb3b7a2/sympy/core/evalf.py#L1427

There are several possible fixes:

1. Make both evalf methods have the same signature (the correct fix)
2. Define n as a function rather than using assignment
3. Add a typehint for n
4. Use `type:ignore`

The reason 2. probably works is that it is (apparently) okay for a subclass to redefine a superclass method and change the signature. However assigning an object of one type to replace/shadow one of another type seems to be an error.

In general muliple inheritance is something that leads to complaints from mypy in the absence of type-hints. If you do:
```python
class A:
    def f(self):
        pass

class B:
    def f(self, x):
        return 2

class C(A, B):
    pass
```
Then you get:
```console
$ mypy t.py
t.py:9: error: Definition of "f" in base class "A" is incompatible with definition in base class "B"
Found 1 error in 1 file (checked 1 source file)
```
However this is okay apparently:
```python
class A:
    def f(self):
        pass

class B:
    def f(self):
        pass

class C(A, B):
    def f(self, x):
        return 2
```
I'm not 100% how this all works but what you need to understand is that we're looking at the errors where there are no type hints and mypy gets confused just trying to analyse the code because there are so many examples of bad practice in the codebase. It will work more reliably when there are lots of type hints and when basic problems like this have been ironed out.

--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -2295,7 +2295,7 @@ class Subs(Expr):
     def evalf(self, prec=None, **options):
         return self.doit().evalf(prec, **options)
 
-    n = evalf
+    n = evalf  # type:ignore
 
     @property
     def variables(self):

